swim/self_evict.go

```diff
@@ -0,0 +1,244 @@
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/ringpop-go/util"
+
+	"github.com/uber-common/bark"
+)
+
+var (
+	// ErrDuplicateHook is returned when a hook with the given name has already
+	// been registered.
+	ErrDuplicateHook = errors.New("expected hook name to be unique")
+
+	// ErrEvictionInProgress is returned when ringpop is already in the process
+	// of evicting
+	ErrEvictionInProgress = errors.New("ringpop is already executing an eviction")
+)
+
+// SelfEvict defines the functions that interact with the self eviction of nodes
+// from the membership prior to shutting down
+type SelfEvict interface {
+	RegisterSelfEvictHook(hooks SelfEvictHook) error
+	SelfEvict() error
+}
+
+// SelfEvictHook is an interface describing a module that can be registered to
+// the self eviction hooks
+type SelfEvictHook interface {
+	// Name returns the name under which the eviction should be regitered
+	Name() string
+
+	// PreEvict is the hook that will be called before ringpop evicts itself
+	// from the membership
+	PreEvict()
+
+	// PostEvict is the hook that will be called after ringpop has evicted
+	// itsels from them memership
+	PostEvict()
+}
+
+type hookFn func(SelfEvictHook)
+
+// SelfEvictOptions configures how self eviction should behave. Applications can
+// configure if ringpop should proactively ping members of the network on self
+// eviction and what percentage/ratio of the memberlist should be pinged at most
+type SelfEvictOptions struct {
+	PingDisable bool
+	PingRatio   float64
+}
+
+func mergeSelfEvictOptions(opt, def SelfEvictOptions) SelfEvictOptions {
+	return SelfEvictOptions{
+		PingDisable: util.SelectBool(opt.PingDisable, def.PingDisable),
+		PingRatio:   util.SelectFloat(opt.PingRatio, def.PingRatio),
+	}
+}
+
+type evictionPhase int
+
+const (
+	preEvict evictionPhase = iota
+	evicting
+	postEvict
+	done
+)
+
+type phase struct {
+	phase evictionPhase
+	start time.Time
+	end   time.Time
+
+	// phase specific information
+	// phase: evicting
+	numberOfPings           int
+	numberOfSuccessfulPings int32
+}
+
+type selfEvict struct {
+	node    *Node
+	options SelfEvictOptions
+	logger  bark.Logger
+	phases  []*phase
+
+	hooks map[string]SelfEvictHook
+}
+
+func newSelfEvict(node *Node, options SelfEvictOptions) *selfEvict {
+	return &selfEvict{
+		node:    node,
+		options: options,
+		logger:  node.logger,
+
+		hooks: make(map[string]SelfEvictHook),
+	}
+}
+
+// RegisterSelfEvictHook registers a named pre/post eviction hook pair.
+func (s *selfEvict) RegisterSelfEvictHook(hook SelfEvictHook) error {
+	name := hook.Name()
+
+	_, hasHook := s.hooks[name]
+	if hasHook {
+		return ErrDuplicateHook
+	}
+
+	s.hooks[name] = hook
+
+	return nil
+}
+
+func (s *selfEvict) SelfEvict() error {
+	if s.currentPhase() != nil {
+		return ErrEvictionInProgress
+	}
+
+	s.logger.Info("ringpop is initiating self eviction sequence")
+
+	// this starts the shutdown sequence, when it returns the complete sequence
+	// has been executed
+	s.preEvict()
```

Comment:
I was thinking the same, but for ease of implementation I wrote it in the exact same way as yours. I will consolidate this into 1 function since it is easier to read in golang than it would have been in nodejs.